### ISSUE TITLE
Compile error due to egm generated files not placed in proto folder

### ISSUE THREAD

**Issue by AlbertoRemus** [iit-danieli-joint-lab/abb_libegm#2]
When compiling the software integrated with the `abb_libegm` library, this error arises:
```
egm.pb.cc
egm_wrapper.pb.cc
egm_wrapper_trajectory.pb.cc
Generating Code...
C:\Users\aremus\aremus-devel\vcpkg-2019-09\vcpkg\installed\x64-windows\include\google/protobuf/descriptor.h(1865): message :  see declaration of 'std::set<std::string,std::less<_Kty>,std::allocator<_Ty>>'
C:\Users\aremus\aremus-devel\vcpkg-2019-09\vcpkg\installed\x64-windows\include\google/protobuf/descriptor.h(1865): message :         with
C:\Users\aremus\aremus-devel\vcpkg-2019-09\vcpkg\installed\x64-windows\include\google/protobuf/descriptor.h(1865): message :         [
C:\Users\aremus\aremus-devel\vcpkg-2019-09\vcpkg\installed\x64-windows\include\google/protobuf/descriptor.h(1865): message :             _Kty=std::string,
C:\Users\aremus\aremus-devel\vcpkg-2019-09\vcpkg\installed\x64-windows\include\google/protobuf/descriptor.h(1865): message :             _Ty=std::string
C:\Users\aremus\aremus-devel\vcpkg-2019-09\vcpkg\installed\x64-windows\include\google/protobuf/descriptor.h(1865): message :         ]
C:\Users\aremus\aremus-devel\vcpkg-2019-09\vcpkg\installed\x64-windows\include\google/protobuf/descriptor.h(1865): message : Please define _WIN32_WINNT or _WIN32_WINDOWS appropriately. For example:
C:\Users\aremus\aremus-devel\vcpkg-2019-09\vcpkg\installed\x64-windows\include\google/protobuf/descriptor.h(1865): message : - add -D_WIN32_WINNT=0x0601 to the compiler command line; or
C:\Users\aremus\aremus-devel\vcpkg-2019-09\vcpkg\installed\x64-windows\include\google/protobuf/descriptor.h(1865): message : - add _WIN32_WINNT=0x0601 to your project's Preprocessor Definitions.
C:\Users\aremus\aremus-devel\vcpkg-2019-09\vcpkg\installed\x64-windows\include\google/protobuf/descriptor.h(1865): message : Assuming _WIN32_WINNT=0x0601 (i.e. Windows 7 target).
C:\Users\aremus\aremus-devel\vcpkg-2019-09\vcpkg\installed\x64-windows\include\google/protobuf/descriptor.h(1865): message : c1xx: fatal error C1083: Cannot open source file: 'C:\Users\aremus\idjl-software\build\_deps\abb_libegm-build\proto\egm.pb.cc': No such file or directory
C:\Users\aremus\aremus-devel\vcpkg-2019-09\vcpkg\installed\x64-windows\include\google/protobuf/descriptor.h(1865): message : c1xx: fatal error C1083: Cannot open source file: 'C:\Users\aremus\idjl-software\build\_deps\abb_libegm-build\proto\egm_wrapper.pb.cc': No such file or directory
C:\Users\aremus\aremus-devel\vcpkg-2019-09\vcpkg\installed\x64-windows\include\google/protobuf/descriptor.h(1865): message : c1xx: fatal error C1083: Cannot open source file: 'C:\Users\aremus\idjl-software\build\_deps\abb_libegm-build\proto\egm_wrapper_trajectory.pb.cc': No such file or directory
```

As it can be noticed at the beginning 3 source files are generated in `~\idjl-software\build\_deps\abb_libegm-build`
```
egm.pb.cc
egm_wrapper.pb.cc
egm_wrapper_trajectory.pb.cc
```
but actually VS compiler complains since it cannot find those files in `~\idjl-software\build\_deps\abb_libegm-build\proto`, so the workaround to solve the problem is to manually place those 3 files in the latter folder, naturally we would like to solve this in an automatic manner

**Comment by traversaro** [comment]
I think this is actually due to https://gitlab.kitware.com/cmake/cmake/issues/18828 . A possible workaround is to move the proto files outside of the `proto` directory and in the main directory of the project, or just overwrite EgmProtoSources with the correct location of the generated files, that I guess is `${CMAKE_CURRENT_BINARY_DIR}/egm.pb.cc` .

**Comment by traversaro** [comment]
Looking a bit more in the original issue https://gitlab.kitware.com/cmake/cmake/issues/18828 , it seems that the problem is just with relative paths of `.proto` files, so it could make sense to try to specify the proto files using absolute files, i.e. substiute: 
~~~
set(EgmProtoFiles proto/egm.proto proto/egm_wrapper.proto proto/egm_wrapper_trajectory.proto)
~~~
with:
~~~
set(EgmProtoFiles ${CMAKE_CURRENT_SOURCE_DIR}/proto/egm.proto
                  ${CMAKE_CURRENT_SOURCE_DIR}/proto/egm_wrapper.proto
                  ${CMAKE_CURRENT_SOURCE_DIR}/proto/egm_wrapper_trajectory.proto)
~~~

**Comment by AlbertoRemus** [comment]
I performed the change here: https://github.com/iit-danieli-joint-lab/abb_libegm/tree/fix-proto-relative-paths-issue

But I had the same problem shown here https://github.com/iit-danieli-joint-lab/abb_libegm/issues/2#issue-528591051

**Comment by AlbertoRemus** [comment]
Of course when compiling I substituted in the main CMakeLists.txt of idjl-software
```
include(FetchContent)
FetchContent_Declare(abb_libegm
                     GIT_REPOSITORY https://github.com/iit-danieli-joint-lab/abb_libegm
                     GIT_TAG 036ca4b7009bc4b441cf7ce1dab14e5f365a1032)
```

where `036ca4b7009bc4b441cf7ce1dab14e5f365a1032` is the tag of the commit in which I did the fix

**Comment by traversaro** [comment]
> I performed the change here: https://github.com/iit-danieli-joint-lab/abb_libegm/tree/fix-proto-relative-paths-issue
> 
> But I had the same problem shown here [#2 (comment)](https://github.com/iit-danieli-joint-lab/abb_libegm/issues/2#issue-528591051)

Can you  try then the workaround mentioned in https://github.com/iit-danieli-joint-lab/abb_libegm/issues/2#issuecomment-558542730 , i.e. moving the `.proto` files outside of the `proto` directory?

**Comment by traversaro** [comment]
By the way, you are using protobuf 3.10, right? While with protobuf 3.6 everything works fine?

**Comment by AlbertoRemus** [comment]
3.9

**Comment by AlbertoRemus** [comment]
> Can you try then the workaround mentioned in #2 (comment) , i.e. moving the .proto files outside of the proto directory?

It failed, but if egm.pb.cc and friends are moved it always works

**Comment by traversaro** [comment]
> > Can you try then the workaround mentioned in #2 (comment) , i.e. moving the .proto files outside of the proto directory?
> 
> It failed, but if egm.pb.cc and friends are moved it always works

Are you sure that you re-configured CMake correctly after  the change? It is quite surprising that moving the `.proto` outside of the `proto` directory, the `.pb.cc` are still searched/placed in the wrong location. Can you point to the commit that contains this workaround? After moving the `.proto` files, did you modified accordingly the `CMakeLists.txt` file?

**Comment by AlbertoRemus** [comment]
https://github.com/iit-danieli-joint-lab/abb_libegm/commit/036ca4b7009bc4b441cf7ce1dab14e5f365a1032

here we are

**Comment by AlbertoRemus** [comment]
but if I have to put `.proto` files outside `proto` folder I shoud do this right?
```
set(EgmProtoFiles ${CMAKE_CURRENT_SOURCE_DIR}/egm.proto
                  ${CMAKE_CURRENT_SOURCE_DIR}/egm_wrapper.proto
                  ${CMAKE_CURRENT_SOURCE_DIR}/egm_wrapper_trajectory.proto)
```

**Comment by traversaro** [comment]
Yes, that is what I was referring to in 

> but if I have to put `.proto` files outside `proto` folder I shoud do this right?
> 
> ```
> set(EgmProtoFiles ${CMAKE_CURRENT_SOURCE_DIR}/egm.proto
>                   ${CMAKE_CURRENT_SOURCE_DIR}/egm_wrapper.proto
>                   ${CMAKE_CURRENT_SOURCE_DIR}/egm_wrapper_trajectory.proto)
> ```

Yes, this is what I was referring to in https://github.com/iit-danieli-joint-lab/abb_libegm/issues/2#issuecomment-559712195 .

**Comment by AlbertoRemus** [comment]
I tried but with this modification the files
```
egm.pb.cc
egm_wrapper.pb.cc
egm_wrapper_trajectory.pb.cc
```
are not generated

**Comment by traversaro** [comment]
Did you re-run CMake or tried with a clean build (just removing `_deps/abb_libegm-*` directories would also probably suffice )

**Comment by AlbertoRemus** [comment]
It was already a clean build

**Comment by traversaro** [comment]
If the commit is https://github.com/ros-industrial/abb_libegm/commit/6c81c73656ce3bc23363a5392c815165dc91d7cb , you also need to move the `.proto` files outside of the `proto` directory, otherwise the `${CMAKE_CURRENT_SOURCE_DIR}/egm.proto` does not exist.

**Comment by AlbertoRemus** [comment]
Got it! Yes indeed it compiles correctly in this case, but still now we have to automatically place protofiles outside protofolder right?

**Comment by traversaro** [comment]
> Got it! Yes indeed it compiles correctly in this case, but still now we have to automatically place protofiles outside protofolder right?

No, those are part of the `abb_libegm` repo, so you can move them by removing  (`git rm`) and adding them back (`git add`) in their new location, and just commit these changes to the repo.
